File: dsa_leetcode_beginner/0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
#include <queue>
#include <vector>

class MedianFinder {
private:
    std::priority_queue<int> leftHalf; // Max-heap
    std::priority_queue<int, std::vector<int>, std::greater<int>> rightHalf; // Min-heap

public:
    MedianFinder() {
    }

    void addNum(int num) {
        if (leftHalf.empty() || num <= leftHalf.top()) {
            leftHalf.push(num);
        } else {
            rightHalf.push(num);
        }

        // Rebalance heaps
        if (leftHalf.size() > rightHalf.size() + 1) {
            rightHalf.push(leftHalf.top());
            leftHalf.pop();
        } else if (rightHalf.size() > leftHalf.size()) {
            leftHalf.push(rightHalf.top());
            rightHalf.pop();
        }
    }

    double findMedian() {
        if (leftHalf.size() > rightHalf.size()) {
            return leftHalf.top();
        } else if (rightHalf.size() > leftHalf.size()) {
            return rightHalf.top();
        } else {
            return (leftHalf.top() + rightHalf.top()) / 2.0;
        }
    }
};

// Usage:
// MedianFinder* obj = new MedianFinder();
// obj->addNum(num);
// double median = obj->findMedian();

File: dsa_leetcode_beginner/0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
import java.util.Collections;
import java.util.PriorityQueue;

class MedianFinder {
    private PriorityQueue<Integer> leftHalf;
    private PriorityQueue<Integer> rightHalf;

    public MedianFinder() {
        leftHalf = new PriorityQueue<>(Collections.reverseOrder()); // Max-heap
        rightHalf = new PriorityQueue<>(); // Min-heap
    }

    public void addNum(int num) {
        if (leftHalf.isEmpty() || num <= leftHalf.peek()) {
            leftHalf.offer(num);
        } else {
            rightHalf.offer(num);
        }

        // Rebalance heaps
        if (leftHalf.size() > rightHalf.size() + 1) {
            rightHalf.offer(leftHalf.poll());
        } else if (rightHalf.size() > leftHalf.size()) {
            leftHalf.offer(rightHalf.poll());
        }
    }

    public double findMedian() {
        if (leftHalf.size() > rightHalf.size()) {
            return leftHalf.peek();
        } else if (rightHalf.size() > leftHalf.size()) {
            return rightHalf.peek();
        } else {
            return (leftHalf.peek() + rightHalf.peek()) / 2.0;
        }
    }
}

// Usage:
// MedianFinder obj = new MedianFinder();
// obj.addNum(num);
// double median = obj.findMedian();
